### PR TITLE
Feature/issue 197 non interactive genesis generator

### PR DESCRIPTION
**Pull request by jgimeno**
This command will be responsable to generate genesis.json files in a non-interactive way to integrate in the CI.

Important changes:
- [x] Something worth noting for reviewers.

Mandatory:
- [x] Added documentation.

Closes #

### CODE REVIEW

**Comment by jgimeno** [comment]
Don't be too bad on my first review!! ;PP just joking. 

I want to discuss some structure ideas about the genesis logic, I thought about setting the command and the handler in its own package, I was thinking about putting it in a folder called genesis/ in the root of the project, so it can look something like genesis.CreateGenesisCommand and it will be opened for everyone if they need to create it's own.

File: cmd/genesis/values.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package main
+
+import (
+	"errors"
+)
+
+const (
+	DefaultSmartContractsOwner = "0x259be75d96876f2ada3d202722523e9cd4dd917d"
```
Comment (yourheropaul) [inline cmd/genesis/values.go]: This value is derived from the [network code](https://github.com/kowala-tech/kcoin/blob/dev/contracts/network/election.go#L20), and I suggest that it only be defined only in election.go. 

In order to do that, `mapChainIDToAddr` would have to be exported and this const would have to refer to it (like `network.MapChainIDToAddr[params.TestnetChainConfig.ChainID.Uint64()]`, which is slightly ugly) but then at least the defaults will be always be up to date with the core code.

File: cmd/genesis/command.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+	ErrEmptyUnbondingPeriod                         = errors.New("unbonding period in days is mandatory")
+	ErrEmptyWalletAddressValidator                  = errors.New("Wallet address of genesis validator is mandatory")
+	ErrInvalidWalletAddressValidator                = errors.New("Wallet address of genesis validator is invalid")
+	ErrEmptyPrefundedAccounts                       = errors.New("empty prefunded accounts, at least the validator wallet address should be included")
+	ErrWalletAddressValidatorNotInPrefundedAccounts = errors.New("prefunded accounts should include genesis validator account")
+	ErrInvalidAddressInPrefundedAccounts            = errors.New("address in prefunded accounts is invalid")
+	ErrInvalidContractsOwnerAddress                 = errors.New("address used for smart contracts is invalid")
+)
+
+type GenerateGenesisCommand struct {
```
Comment (yourheropaul) [inline cmd/genesis/command.go]: Why make this type exportable with only non-exported members? Maybe I'm misunderstanding the use cases for it, but shouldn't it be either an interface or have at least one exported member?

File: cmd/genesis/command.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	}
+
+	// Add a batch of precompile balances to avoid them getting deleted
+	for i := int64(0); i < 256; i++ {
+		genesis.Alloc[common.BigToAddress(big.NewInt(i))] = core.GenesisAccount{Balance: big.NewInt(1)}
+	}
+
+	out, _ := json.MarshalIndent(genesis, "", "  ")
+
+	_, err = h.w.Write(out)
```
Comment (yourheropaul) [inline cmd/genesis/command.go]: This comment is quite anal, but can the conditional not be one line here?

File: cmd/genesis/vmtracer.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	data map[common.Address]map[common.Hash]common.Hash
+}
+
+func newVmTracer() *vmTracer {
+	return &vmTracer{
+		data: make(map[common.Address]map[common.Hash]common.Hash, 1024),
+	}
+}
+
+func (vmt *vmTracer) CaptureState(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, memory *vm.Memory, stack *vm.Stack, contract *vm.Contract, depth int, err error) error {
```
Comment (yourheropaul) [inline cmd/genesis/vmtracer.go]: I know this is copied from puppeth, but without tests this function is very confusing. What does it do?

Comment (jgimeno) [reply cmd/genesis/vmtracer.go]: I think this is a tracer to capture some kind of information from the real blockchain, is there anybody that can assert this info?

File: cmd/genesis/values.go (hunk 1 above)
Comment (jgimeno) [reply cmd/genesis/values.go]: Great! Will do this.

Comment (jgimeno) [reply cmd/genesis/values.go]: I understand that this should be defined based on the network selection, main or test

File: kcoin/genesis/genesis_generator.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+
+func setPrefundedAccounts(validPrefundedAccounts []*validPrefundedAccount, genesis *core.Genesis) {
+	for _, vAccount := range validPrefundedAccounts {
+		genesis.Alloc[*vAccount.walletAddress] = core.GenesisAccount{
+			Balance: vAccount.balance,
+		}
+	}
+}
+
+func createNetwork(network string) (string, error) {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator.go]: `get`, `map` or similar on these func names, I think `create` implies creating a object or a resource, this just maps between 2 values

File: kcoin/genesis/genesis_generator.go [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+	if !availableNetworks[network] {
+		return "", ErrInvalidNetwork
+	}
+
+	return network, nil
+}
+
+func createConsensusEngine(consensus string) (string, error) {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator.go]: same as `createNetwork`

File: kcoin/genesis/genesis_generator.go [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+func createUnbondingPeriod(uP string) (*big.Int, error) {
+	var text string
+	if text = strings.TrimSpace(uP); text == "" {
+		return nil, ErrEmptyUnbondingPeriod
+	}
+
+	unbondingPeriod, ok := new(big.Int).SetString(text, 0)
+	if !ok {
+		//TODO: Create error
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator.go]: create the error then :)

File: kcoin/genesis/genesis_generator.go [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+		return nil, errors.New("invalid max num of validators.")
+	}
+
+	return unbondingPeriod, nil
+}
+
+func createWalletAddress(wA string) (*common.Address, error) {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator.go]: `wA` A uppercase on purpose?

File: cmd/genesis/genesis.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/kowala-tech/kcoin/kcoin/genesis"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"os"
+	"strconv"
+	"strings"
+)
+
+var (
+	FileConfig string
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply cmd/genesis/genesis.go]: right, this was the first I did and I left it. Good eye!

File: kcoin/genesis/genesis_generator.go (hunk 5 above)
Comment (jgimeno) [reply kcoin/genesis/genesis_generator.go]: Thanks, I had some issues with the naming on this functions.

File: kcoin/genesis/genesis_generator.go [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+	}
+
+	genesis := &core.Genesis{
+		Timestamp: uint64(time.Now().Unix()),
+		GasLimit:  4700000,
+		Alloc:     make(core.GenesisAlloc),
+		Config:    &params.ChainConfig{},
+	}
+
+	setNetwork(validOptions.network, genesis)
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply kcoin/genesis/genesis_generator.go]: good one!

File: kcoin/genesis/genesis_generator.go [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+	case MainNetwork:
+		genesis.Config.ChainID = params.MainnetChainConfig.ChainID
+	case TestNetwork:
+		genesis.Config.ChainID = params.TestnetChainConfig.ChainID
+	case OtherNetwork:
+		genesis.Config.ChainID = new(big.Int).SetUint64(uint64(rand.Intn(65536)))
+	}
+}
+
+func validateOptions(options GenesisOptions) (*validGenesisOptions, error) {
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply kcoin/genesis/genesis_generator.go]: umm, I have my doubts in this one, since finally for me it validates if the options are correct, and if they are they return a validOptions struct. Can it be?

[the remark this answers is not in the document]
Comment (jmatosp) [reply kcoin/genesis/genesis_generator.go]: :) nw

does both things, depends on what you are seeing as main, does validate and map to a diff struct

You can leave as is

File: cmd/genesis/genesis.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func main() {
+	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func parsePrefundedAccounts(accounts interface{}) []genesis.PrefundedAccount {
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply cmd/genesis/genesis.go]: This is needed since data can come from 2 types, string or slice depending on where it is loaded (config or command line). I preferred to have this encapsulated in this function instead of having an if on the callee function. It is not the best, but not really really the worst. If you still want to fix this let me know! :D

File: kcoin/genesis/genesis_generator.go [hunk 13]
```diff
@@ -65,7 +65,7 @@ type GenesisOptions struct {
 
 type PrefundedAccount struct {
 	WalletAddress string
-	Balance       int64
+	Balance       string
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator.go]: why moving to string?

File: cmd/genesis/genesis.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	prefundedAccounts := make([]genesis.PrefundedAccount, 0)
+
+	switch accounts.(type) {
+	// Accounts come from config file.
+	case []interface{}:
+		accountArray := accounts.([]interface{})
```
Comment (jmatosp) [inline cmd/genesis/genesis.go]: ah, now I get it, maybe the comment wont be necessary if you extract each block of the case to it's own function

`prefundedAccountsFromConfig` and `prefundedAccountsFromCLI`?

```
switch accounts.(type) {
   case []interface{}:
      accountArray := accounts.([]interface{})
      prefundedAccountsFromConfig(accountArray)

   case string:
      accountsString := accounts.(string)
      prefundedAccountsFromCLI(accountsString)
}
```

File: kcoin/genesis/genesis_generator.go (hunk 13 above)
Comment (jgimeno) [reply kcoin/genesis/genesis_generator.go]: I was having problems to manage big numbers from the console or the config files. All the values in the options are of type string, in this case we can benefit too different types of numbers for the configuration as decimal, hex, etc.

After it is converted to big int, address, etc

File: kcoin/genesis/genesis_generator_test.go [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+				return command
+			},
+			ExpectedError: ErrInvalidConsensusEngine,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.TestName, func(t *testing.T) {
+			_, err := GenerateGenesis(test.InvalidCommandFromValid(baseValidCommand))
+			if err != test.ExpectedError {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: `assert.EqualError(t, test.ExceptedError, err)`

I think this will log the case under test

File: kcoin/genesis/genesis_generator_test.go [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+		PrefundedAccounts: []PrefundedAccount{
+			{
+				WalletAddress: "0xe2ac86cbae1bbbb47d157516d334e70859a1bee4",
+				Balance:       "15",
+			},
+		},
+	}
+
+	generatedGenesis, err := GenerateGenesis(cmd)
+
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: `assert.NoError(t, err)`

File: kcoin/genesis/genesis_generator_test.go [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+	if err != nil {
+		t.Fatalf("Error: %s", err.Error())
+	}
+
+	fileName := "testfiles/testnet_default.json"
+	contents, err := ioutil.ReadFile(fileName)
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: `assert.NoError(t, err)` or `require.NoError(t, err)`

File: kcoin/genesis/genesis_generator_test.go [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+	if err != nil {
+		t.Fatalf("Failed to read file %s", fileName)
+	}
+
+	var expectedGenesis = new(core.Genesis)
+	err = json.Unmarshal(contents, expectedGenesis)
+	if err != nil {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: assert.NoError(t, err) or require.NoError(t, err)

File: kcoin/genesis/genesis_generator_test.go [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+		},
+	}
+
+	generatedGenesis, err := GenerateGenesis(cmd)
+
+	if err != nil {
+		t.Fatalf("Error: %s", err.Error())
+	}
+
+	fileName := "testfiles/testnet_default.json"
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: yeah! golden files!

add a 

`var update = flag.Bool("update", false, "update .golden files")`

and then

```
if *update {
		t.Log("update golden file")
		if err := ioutil.WriteFile(gp, b.Bytes(), 0644); err != nil {
			t.Fatalf("failed to update golden file: %s", err)
		}
	}
```

so you can generate the golden file from tests when flag is set

`go test -v ./... -update`

more details
https://medium.com/soon-london/testing-with-golden-files-in-go-7fccc71c43d3

File: kcoin/genesis/genesis_generator_test.go [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+		}
+
+		expectedExtradata := make([]byte, 32)
+		expectedExtradata = append([]byte(extraDataStr), expectedExtradata[len(extraDataStr):]...)
+
+		assert.Equal(t, expectedExtradata, generatedGenesis.ExtraData)
+	})
+}
+
+func TestItFailsWithAnInvalidNetwork(t *testing.T) {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: isn't this a dup test from the beginning of the file, or just the name `TestMapNetwork`?

File: kcoin/genesis/genesis_generator_test.go [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+		t.Run(test.testName, func(t *testing.T) {
+			_, err := mapNetwork(test.network)
+			if err != ErrInvalidNetwork {
+				t.Fatalf("Failed to throw exception with an invalid Network value.")
+			}
+		})
+	}
+}
+
+func TestItFailsWithInvalidConsensusEngine(t *testing.T) {
```
Comment (jmatosp) [inline kcoin/genesis/genesis_generator_test.go]: same here `TestMapConsensusEngine`?

File: kcoin/genesis/genesis_generator.go (hunk 13 above)
Comment (jmatosp) [reply kcoin/genesis/genesis_generator.go]: :+1:

File: cmd/genesis/genesis.go (hunk 14 above)
Comment (jgimeno) [reply cmd/genesis/genesis.go]: nice

File: cluster/genesis.go [hunk 22]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
 	if err != nil {
 		return err
 	}
-	var b bytes.Buffer
-	genesisTpl.Execute(&b, map[string]string{
-		"NetworkID": client.NetworkID,
-	})
+
+	newGenesis, err := genesis.GenerateGenesis(
+		genesis.GenesisOptions{
```
Comment (rgeraldes) [inline cluster/genesis.go]: genesis.Options vs genesis.GenesisOptions?

File: cluster/genesis.go [hunk 23]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
+			Network:                       "test",
+			MaxNumValidators:              "1",
+			UnbondingPeriod:               "0",
+			WalletAddressGenesisValidator: "0xd6e579085c82329c89fca7a9f012be59028ed53f",
```
Comment (rgeraldes) [inline cluster/genesis.go]: AccountAddressGenesisValidator

File: cluster/genesis.go [hunk 24]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
+			PrefundedAccounts: []genesis.PrefundedAccount{
+				{
+					WalletAddress: "0xd6e579085c82329c89fca7a9f012be59028ed53f",
```
Comment (rgeraldes) [inline cluster/genesis.go]: AccountAddress

File: cluster/genesis.go [hunk 25]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
+					Balance:       "0x200000000000000000000000000000000000000000000000000000000000000",
+				},
+				{
+					WalletAddress: "0x259be75d96876f2ada3d202722523e9cd4dd917d",
+					Balance:       "1000000000000000000",
+				},
+			},
+			SmartContractsOwner: "0x259be75d96876f2ada3d202722523e9cd4dd917d",
```
Comment (rgeraldes) [inline cluster/genesis.go]: This is probably going to be multisig: meaning that a contract will probably have multiple keys as the owner - please open an issue for this specific problem. We need to check this! :) For now we can leave it as is.

Comment (rgeraldes) [reply cluster/genesis.go]: SmartContractsOwner is too broad. I'd call it CoreContractsOwner

File: cluster/genesis.go [hunk 26]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
+		},
+	)
+
```
Comment (rgeraldes) [inline cluster/genesis.go]: extra space

File: cluster/minikube.go [hunk 27]
```diff
@@ -77,7 +77,7 @@ func (cluster *minikubeCluster) Delete() error {
 
 // DockerEnv returns the environment variables necessary to connect to the private docker repository in the kubernetes cluster
 func (cluster *minikubeCluster) DockerEnv() ([]string, error) {
-	statusCmd := exec.Command("minikube", "docker-env", "-p", cluster.Name)
+	statusCmd := exec.Command("minikube", "docker-env", "-p", cluster.Name, "--shell", "bash")
```
Comment (rgeraldes) [inline cluster/minikube.go]: Had a quick look at the file and it seems that we are not checking errors for some cmd executions - not sure if its expected.

**Comment by rgeraldes** [comment]
good job!

File: cluster/minikube.go (hunk 27 above)
Comment (rgeraldes) [reply cluster/minikube.go]: Other: Also the internal/build/utils.go includes some utils for cmds. I'm not sure if its the best place though

Comment (acroca) [reply cluster/minikube.go]: @rgeraldes In all cmd executions we check for it's exit status to be 0. Where did you find a cmd without error checks?

File: cmd/genesis/genesis.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (acroca) [reply cmd/genesis/genesis.go]: I'd also prefer to either be specific with the data we're sending (if it's a string, use a separate func than the slice), or use always slices, even if there's only one element. Is that possible?

File: kcoin/genesis/genesis_generator_test.go [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"math/big"
+	"path/filepath"
+	"testing"
+)
+
+var update = flag.Bool("update", false, "update .golden files")
+
+func TestItFailsWhenRunningHandlerWithInvalidCommandValues(t *testing.T) {
```
Comment (acroca) [inline kcoin/genesis/genesis_generator_test.go]: Just a personal preference, I like to see positive tests first, errors/edge cases later. Main reason is to open a test file and the first thing to see is a kind of example of how this works, if you need more detail you can read more.

File: cluster/genesis.go [hunk 29]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
 		return err
 	}
-	var b bytes.Buffer
-	genesisTpl.Execute(&b, map[string]string{
-		"NetworkID": client.NetworkID,
-	})
+
+	newGenesis, err := genesis.GenerateGenesis(
+		genesis.GenesisOptions{
+			Network:                       "test",
```
Comment (acroca) [inline cluster/genesis.go]: You're not using the `client.NetworkID` anymore. We should change it to `client.Network` and use it here.

File: cluster/genesis.go [hunk 30]
```diff
@@ -836,10 +21,32 @@ func (client *cluster) generateGenesis() error {
...
+			},
+			SmartContractsOwner: "0x259be75d96876f2ada3d202722523e9cd4dd917d",
+		},
+	)
+
+	if err != nil {
+		return err
+	}
+
+	out, _ := json.MarshalIndent(newGenesis, "", "  ")
```
Comment (acroca) [inline cluster/genesis.go]: Handle the error

File: kcoin/genesis/genesis_generator_test.go (hunk 28 above)
Comment (jgimeno) [reply kcoin/genesis/genesis_generator_test.go]: Interesting point of view, never thought about it, has sense. I normally start with the edge cases than the happy path, that is why it appears in this order. I will change it.

File: cmd/genesis/genesis.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (acroca) [reply cmd/genesis/genesis.go]: if it overcomplicates things, then I'm fine with this. @jmatosp?

**Comment by jgimeno** [comment]
@rgeraldes we include network in this case, that can be test or main. Do you want a specific   NetworkID/ChainID too? Is it needed?
